hypershift-operator/controllers/platform/aws/controller.go

```diff
@@ -523,43 +523,53 @@ func (r *AWSEndpointServiceReconciler) delete(ctx context.Context, awsEndpointSe
 		ServiceIds: []*string{aws.String(serviceID)},
 	})
 	if err != nil {
-		log.Error(err, "Failed to delete vpc endpoint service configuration", "serviceID", serviceID)
-		existingConnectionsResult, describeConnectionsErr := r.ec2Client.DescribeVpcEndpointConnectionsWithContext(ctx, &ec2.DescribeVpcEndpointConnectionsInput{
-			Filters: []*ec2.Filter{
-				{
-					Name:   aws.String("service-id"),
-					Values: []*string{aws.String(serviceID)},
-				},
-			},
-		})
-		if describeConnectionsErr != nil {
-			log.Error(describeConnectionsErr, "Failed to get existing connections", "serviceID", serviceID)
-			return false, unwrapError(describeConnectionsErr)
-		}
-		var existingEndpointIDs []*string
-		for _, conn := range existingConnectionsResult.VpcEndpointConnections {
-			state := aws.StringValue(conn.VpcEndpointState)
-			switch state {
-			case "pendingAcceptance", "pending", "available":
-				existingEndpointIDs = append(existingEndpointIDs, conn.VpcEndpointId)
-			}
-		}
-		if len(existingEndpointIDs) > 0 {
-			if _, rejectEndpointsErr := r.ec2Client.RejectVpcEndpointConnectionsWithContext(ctx, &ec2.RejectVpcEndpointConnectionsInput{
-				ServiceId:      aws.String(serviceID),
-				VpcEndpointIds: existingEndpointIDs,
-			}); rejectEndpointsErr != nil {
-				return false, unwrapError(rejectEndpointsErr)
-			}
-		}
 		return false, unwrapError(err)
 	}
+
+	// DeleteVpcEndpointServiceConfigurations doesn't return errors directly when a VPC Endpoint Service doesn't exist
+	// or when it has active connections, instead returning errors within output.Unsuccessful
+	// https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_DeleteVpcEndpointServiceConfigurations.html
 	if output != nil && len(output.Unsuccessful) != 0 && output.Unsuccessful[0].Error != nil {
+		log.Error(err, "unsuccessful deleting vpc endpoint service", "serviceID", serviceID)
```

Comment:
@mjlshen can we turn this error handling code into a function and call it from both the actual error case *and* when the error is returned in the response?

Reply:
Sure thing! Just for my understanding, are you suggesting this to kind of "future-proof" this code in case AWS does the right thing and returns the "actual error" in the future?

Reply:
I took a bit of journey to make the e2e tests runnable locally again in https://github.com/openshift/hypershift/pull/2742 and plan to get this logic + tested in e2e.

Reply:
This isn't very trivial as the actual error satisfies the `awserr.Error` interface, but `ec2.UnsuccessfulItemError` does not - and doesn't even satisfy the general Go `Error` interface - https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#UnsuccessfulItemError.

But after experimenting with this by hand, I have found that `InvalidVpcEndpointService.NotFound` and `InvalidVpcEndpointServiceId.NotFound` only get returned as UnsuccessfulItemErrors when `ec2:DeleteVpcEndpointServiceConfigurations` is used.